Node id assignment

andrwng opened the pull request:
## Cover letter

Describe in plain language the motivation (bug, feature, etc.) behind the change in this PR and how the included commits address it.

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [ ] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

Describe in plain language how this PR affects an end-user. What topic flags, configuration flags, command line flags, deprecation policies etc are added/changed.

<!-- don't ship user breaking changes. Ping PMs for help with user visible changes  -->

## Release notes
<!--

If this PR does not need to be included in the release notes, then
simply have a bullet point for `none` directly under the `Release notes`
section, e.g.

* none

Otherwise, add one or more of the following sections. A section must have
at least 1 bullet point. You can add multiple sections with multiple
bullet points if this PR represents multiple release note items. See
the CONTRIBUTING.md guidelines for more details.

### Features

* Short description of the feature. Explain how to configure the new feature if applicable.

### Improvements

* Short description of how this PR improves redpanda.

-->

andrwng committed:
storage: add node_uuid to storage API

This will be used when we first start a node that has no node UUID.
I exposed this as storage because the UUID uniquely identifies the
contents of the kv-store / data directory.

andrwng committed:
cluster: introduce bootstrap service

This adds the skeleton of an RPC service that will be used for cluster
bootstrap. It includes an initial `cluster_bootstrap_info` RPC type that
is currently empty.

Introducing as a standalone commit to parallelize some workstreams.

andrwng committed:
cluster: encapsulate cluster discovery

This encapsulates some pieces of cluster discovery into its own class. I
added some descriptions on how each method is expected to be
implemented, but I added implementations that match the existing
behavior.

andrwng committed:
application: generate node UUID on startup

Upon first starting up the storage layer, Redpanda will now check to see
if a node UUID exists in the controller keyspace of the key-value store.
If not, it will generate one and persist it to disk, and make it
available on all shards.

This will be used as a unique identifier when registering a node
(particularly important in the future when we auto-assign node IDs).

dlex commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -0,0 +1,50 @@
+namespace cluster {
+
+cluster_discovery::cluster_discovery(
+  const config::node_config& node_config, const model::node_uuid& node_uuid)
```
It's an object that is availiable statically, probably no need for passing as an argument

andrwng replied:
Removed

dlex commented on src/v/cluster/members_manager.h:
```diff
@@ -127,21 +186,55 @@ class members_manager {
+    // get_node_updates().
     ss::queue<node_update> _update_queue;
+
+    map_t _uuid_by_id;
```
it's vice versa in my mind - id by uuid

andrwng replied:
Done

dlex commented on src/v/serde/test/struct_gen.py:
```diff
@@ -17,6 +17,7 @@ class BasicType(Enum):
     CHAR = 5
     FLOAT = 6
     DOUBLE = 7
+    UUID = 8
```
Missing in `def __str__(self)`:

```
        elif self._basic_type == BasicType.UUID:
            return "uuid_t"
```

andrwng committed:
cluster: command to register a UUID

This adds a new controller command to register a node UUID.

It is activated by the existing join RPC endpoint and has the following
semantics:
- Register: all new versions of Redpanda will send their UUID along with
  an optional node ID in their initial join request. If the UUID is not
  in the leader controller's `members_table`, the UUID is registered by
  replicating a `register_node_uuid_cmd`.
- Join: all new version of Redpanda will, after registering, send
  another join request that will actually add the node to the Raft
  group. Old versions of Redpanda will send a join request with no UUID.

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -30,7 +31,20 @@ cluster_discovery::cluster_discovery(
+        clusterlog.info("Using configured node ID {}", configured_node_id);
+        co_return configured_node_id;
+    }
+    // TODO: once is_cluster_founder() refers to all seeds, verify that all the
+    // seeds' seed_servers lists match and assign node IDs based on the
+    // ordering.
```
cc @dlex

andrwng committed:
net: add node_id label to client metrics

RPC `client_probe` metrics currently leverage a labeling scheme defined
by Seastar in which, for a given metric, a set of metric labels can only
be registered once per metric. Currently, the label used is solely based
on the server address associated with a given `rpc::transport`. As such,
we currently cannot start multiple `rpc::transport`s pointed at the same
server.

This functionality could be useful though: consider when a node is
restarted empty with a new node ID. Redpanda currently has a check that
nodes being added to the controller Raft group don't overlap with
existing group members' addresses. But if we were to remove this check,
when the new node _is_ added to the Raft group, each node will try to
create a new `rpc::transport` pointing at the new node, and register
metrics with identical labels to those registered by the old node, and
be met with a Seastar `double_registration` exception.

To enable the above scenario, this commit adds the `node_id` as a label
for client metrics, and aggregates them by this label.

andrwng committed:
features: add flag for node_id_assignment

andrwng committed:
cluster: allow node to join if address is used

andrwng committed:
tests: ducktape test for node ID assignment

dlex commented on src/v/redpanda/tests/fixture.h:
```diff
@@ -62,25 +65,30 @@ class redpanda_thread_fixture {
       std::optional<scheduling_groups> sch_groups,
-      bool remove_on_shutdown)
+      bool remove_on_shutdown,
+      configure_node_id use_node_id = configure_node_id::yes)
```
If instead the `node_id` parameter was made `std::optional<model::node_id>`, that would have saved an argument, and also avoided ambiguity of `node_id` value when `configure_node_id::no`.

andrwng replied:
We chatted about this -- it ended up being necessary to differentiate between different nodes (e.g. differing ports, data dirs, logger names, etc). Added a comment explaining this

andrwng committed:
util: add named_type static constructor from args

It can be convenient to construct a named_type using constructors of the
underlying type.

A simple solution today is to construct the underlying type separately
and then construct the named_type with it, but this spills the
underlying type into call sites. Another approach would be to construct
with the named_type::type constructor, but this can be verbose.

This commit adds a templatized static constructor that callers can pass
constructor args into.

andrwng committed:
config: make node ID optional

This commit makes the node_id config property optional. It doesn't make
any checks that the field is set. A follow-up will change the Redpanda's
behavior when seeing an empty node ID.

andrwng committed:
cluster: unit tests for node ID assignment

andrwng committed:
cluster: auto-assign node ID when empty

This commit adds the ability for Redpanda to assign node ID based on the
node UUID when a node ID isn't set.

andrwng committed:
tests: add option to not configure node ID

andrwng committed:
cluster_discovery: read node ID from kvstore

When the controller starts up, we check to ensure the configured node ID
is identical to that in the kvstore if one exists, persisting one if
not. We should use this node ID at startup if weren't configured with
one, e.g. if we're upgrading an existing cluster and begin omitting node
ID from the node config.

andrwng committed:
cluster_discovery: add an abort source

andrwng committed:
redpanda: plumb cluster discovery into application

This plumbs the cluster discovery module into `redpanda::application` as
one of the first subsystems initialized during startup.

Cluster discovery will have an initial dependency on the storage
subsystem, to get node UUID, cluster UUID, etc., and on the RPC
subsystem, to serve the RPCs required for the initial bootstrapping
sequence.

The rest of the subsystems will have a dependency on cluster discovery,
which, once automatic node ID assignment is supported, will be what
determines the node ID.

To that end, this commit orders the startup of the storage and RPC
subsystems up front, followed by cluster discovery, followed by the rest
of the subsystems.

This commit only plumbs cluster discovery infrastructure. The cluster
discovery sequence as a whole is unchanged.
